input/musig2.go

```diff
@@ -168,11 +325,200 @@ func MuSig2CombineKeys(allSignerPubKeys []*btcec.PublicKey,
+// BIP draft version 1.0.0rc2.
+func createContextV100RC2(privKey *btcec.PrivateKey,
+	allSignerPubKeys []*btcec.PublicKey,
+	tweaks *MuSig2Tweaks) (*musig2.Context, *musig2.Session, error) {
+
+	// The context keeps track of all signing keys and our local key.
+	allOpts := append(
+		[]musig2.ContextOption{
+			musig2.WithKnownSigners(allSignerPubKeys),
+		},
+		tweaks.ToContextOptions()...,
+	)
+	muSigContext, err := musig2.NewContext(privKey, true, allOpts...)
+	if err != nil {
+		return nil, nil, fmt.Errorf("error creating MuSig2 signing "+
+			"context: %v", err)
+	}
+
+	muSigSession, err := muSigContext.NewSession()
+	if err != nil {
+		return nil, nil, fmt.Errorf("error creating MuSig2 signing "+
+			"session: %v", err)
+	}
+
+	return muSigContext, muSigSession, nil
+}
+
+// createContextV040 implements the MuSig2CreateContext logic for the MuSig2 BIP
+// draft version 0.4.0.
+func createContextV040(privKey *btcec.PrivateKey,
+	allSignerPubKeys []*btcec.PublicKey,
+	tweaks *MuSig2Tweaks) (*musig2v040.Context, *musig2v040.Session,
+	error) {
+
+	// The context keeps track of all signing keys and our local key.
+	allOpts := append(
+		[]musig2v040.ContextOption{
+			musig2v040.WithKnownSigners(allSignerPubKeys),
+		},
+		tweaks.ToV040ContextOptions()...,
+	)
+	muSigContext, err := musig2v040.NewContext(privKey, true, allOpts...)
+	if err != nil {
+		return nil, nil, fmt.Errorf("error creating MuSig2 signing "+
+			"context: %v", err)
+	}
+
+	muSigSession, err := muSigContext.NewSession()
+	if err != nil {
+		return nil, nil, fmt.Errorf("error creating MuSig2 signing "+
+			"session: %v", err)
+	}
+
+	return muSigContext, muSigSession, nil
+}
+
+// MuSig2Sign calls the Sign() method on the given versioned signing session and
+// returns the result in the most recent version of the MuSig2 API.
+func MuSig2Sign(session MuSig2Session, msg [32]byte,
+	withSortedKeys bool) (*musig2.PartialSignature, error) {
+
+	switch s := session.(type) {
+	case *musig2.Session:
+		var opts []musig2.SignOption
+		if withSortedKeys {
+			opts = append(opts, musig2.WithSortedKeys())
+		}
+		partialSig, err := s.Sign(msg, opts...)
+		if err != nil {
+			return nil, fmt.Errorf("error signing with local key: "+
+				"%v", err)
+		}
+
+		return partialSig, nil
+
+	case *musig2v040.Session:
+		var opts []musig2v040.SignOption
+		if withSortedKeys {
+			opts = append(opts, musig2v040.WithSortedKeys())
+		}
+		partialSig, err := s.Sign(msg, opts...)
+		if err != nil {
+			return nil, fmt.Errorf("error signing with local key: "+
+				"%v", err)
+		}
+
+		return &musig2.PartialSignature{
+			S: partialSig.S,
+			R: partialSig.R,
+		}, nil
+
+	default:
+		return nil, fmt.Errorf("invalid session type <%T>", s)
+	}
+}
+
+// MuSig2CombineSig calls the CombineSig() method on the given versioned signing
+// session and returns the result in the most recent version of the MuSig2 API.
+func MuSig2CombineSig(session MuSig2Session,
```

Comment:
The interface above https://github.com/lightningnetwork/lnd/blob/4e3decd26baf9b5915f8fb036f53ec73e30c3394/input/musig2.go#L76 has a different signature (return types). Is that intended? Sorry if I'm missing something here

Reply:
I agree, this might be a bit confusing. The `MuSig2Signer` signer interface is the "public" interface that you would present to a user, for example over RPC. So the user only needs to know the session ID and doesn't need to maintain the full session struct.
While on the other hand, the global functions in this package (like the mentioned `MuSig2CombineSig`) are the actual backend implementations that the RPC server would call. So basically a call would go as `User -> RPC server -> backend function`, whereas the RPC server would keep track of the signing sessions in memory.

Reply:
Thanks for the explanation